File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        INT_MAX = 2**31 - 1
        INT_MIN = -2**31
        i = 0
        while i < len(s) and s[i] == ' ':
            i += 1
        sign = 1
        if i < len(s) and (s[i] == "+" or s[i] == '-'):
            if s[i] == '-':
                sign = -1
            i += 1
        
        num = 0
        while i < len(s) and s[i].isdigit():
            digit = int(s[i])
            num = num * 10 + digit

            if num > INT_MAX:
                return INT_MAX if sign == 1 else INT_MIN

            i += 1

        return sign * num

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.cpp
#include <climits>
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    int myAtoi(string s) {
        int i = 0;
        while (i < s.size() && s[i] == ' ')
            i++;

        int sign = 1;
        if (i < s.size() && (s[i] == '+' || s[i] == '-')) {
            if (s[i] == '-')
                sign = -1;
            i++;
        }

        long result = 0;
        while (i < s.size() && isdigit(s[i])) {
            result = result * 10 + (s[i] - '0');

            if (result * sign > INT_MAX)
                return INT_MAX;
            if (result * sign < INT_MIN)
                return INT_MIN;

            i++;
        }

        return (int)(result * sign);
    }
};
